### ISSUE TITLE
LogoutUri and ADFS integration

### ISSUE THREAD

**Issue by solidstore** [IdentityServer/IdentityServer3#3110]
I have local logins and ADFS (WS-Fed) logins enabled in my identity server. We have implemented the signout endpoint and set that in the client LogoutUri. This is working correctly with local users. However, with ADFS users the identity server logout page doesnt render the iframe that would call the callback but instead redirects to the ADFS signout screens.

I'm trying to implement single sign out so that any application using the identity server signs out all the other applications but this doesnt seem possible with ADFS? Or am I missing something?

I noticed in fiddler that ADFS appears to callback to IdServer with a request to /identity/adfs?wa=wsignoutcleanup1.0 but this isnt found.

**Comment by solidstore** [comment]
I can handle the wsignoutcleanup callback inside the identityserver using the following technique:
https://leastprivilege.com/2015/07/08/federated-logout-with-the-katana-ws-federation-middleware/

but I dont seem to be able to figure out how to get my client LogoutUri called by IdentityServer triggered by a callback from ADFS.

**Comment by solidstore** [comment]
OK, I have made some progress. I have now got ADFS calling back to the identity server on signout using the wreply parameter AND with wsignoutcleanup1.0. Key post in case anyone else struggling: https://www.sevecek.com/EnglishPages/Lists/Posts/Post.aspx?ID=93 You **MUST** make the signout callback the default WS-Fed endpoint! 

Q1: I'm not sure what code should be in wsignoutcleanup1.0 handler and what goes in other callback?

I have followed the example here to try to render the loggedout page:
https://identityserver.github.io/Documentation/docsv2/advanced/federated-post-logout-redirect.html

and call RenderLoggedOutViewAsync, but the response just returns an error 404. First it complained about missing Mvc controllers so I added an IgnoreRoutes entry but now I get a different 404.

My breakpoints show that the RenderLoggedOutViewAsync method is run, but i dont see its output in the fiddler traces.

Q2: Any ideas why the logged out page isnt returned in the response?

log for signout:

```
2016-08-03 12:14:42,499 INFO  [22  ] Logging.TraceSourceSink                  End end session request  
2016-08-03 12:14:42,502 INFO  [22  ] Logging.TraceSourceSink                  Redirecting to logout page  
2016-08-03 12:14:42,513 INFO  [21  ] Logging.TraceSourceSink                  Logout prompt for subject: "{752e101f-90bb-44ab-9e3b-16960653e11e}"  
2016-08-03 12:14:42,520 INFO  [12  ] Logging.TraceSourceSink                  SignOutMessage present (from client "0BE46598-C5FF-447F-988D-EA106936BB9E") and RequireSignOutPrompt is false, performing logout  
2016-08-03 12:14:42,525 INFO  [12  ] Logging.TraceSourceSink                  Logout endpoint submitted  
2016-08-03 12:14:42,527 INFO  [12  ] Logging.TraceSourceSink                  Logout requested for subject: "{752e101f-90bb-44ab-9e3b-16960653e11e}"  
2016-08-03 12:14:42,529 INFO  [12  ] Logging.TraceSourceSink                  Clearing cookies  
2016-08-03 12:14:42,536 INFO  [12  ] Logging.TraceSourceSink                  rendering logged out page  

```

requests seen in fiddler (simplified):

```
websvr      /Portal/signout.aspx
websvr      /identity/connect/endsession
websvr     /identity/logout?id=xxxx
adfs         /adfs/ls/?wa=wsignout&wreply=signoutcallback
websvr    /identity/adfs?wa=wsignoutcleanup1.0
websvr    /identity/signoutcallback      <---- 404, even though callback hit and RenderLoggedOutViewAsync called
```

startup:

            var options = new IdentityServerOptions
            {
                ... options removed to simplify

                AuthenticationOptions = new AuthenticationOptions
                {
                     ... options removed to simplify

                    IdentityProviders = ConfigureIdentityProviders
                },
            };

            appBuilder.UseIdentityServer(options);

            appBuilder.Map("/signoutcallback", cleanup =>
            {
                cleanup.Run(async ctx =>
                {
                    var state = ctx.Request.Cookies["state"];
                    await ctx.Environment.RenderLoggedOutViewAsync(state);
                });
            });
        
            appBuilder.Map("/signoutcleanup", cleanup =>
            {
                cleanup.Run(async ctx =>
                {
                    await ctx.Environment.ProcessFederatedSignoutAsync();
                });
            });
        }

        private void ConfigureIdentityProviders(IAppBuilder app, string signInAsType)
        {
               app.Use(async (ctx, next) =>
               {
                    var qs = ctx.Request.Query;
                    var wa = qs.Get("wa");

                    if (wa != null)
                    {
                        if (wa == "wsignoutcleanup1.0")
                        {
                            // not sure what to do here?

                            //await ctx.Environment.ProcessFederatedSignoutAsync();
                            // clean up resources, e.g. the logon cookie
                            //ctx.Authentication.SignOut(); 
                        }
                    }

                    await next();
                });

                // ADFS identity provider
                var adfs = new WsFederationAuthenticationOptions
                {
                        AuthenticationType = AnalyticIdentityProviders.External,
                        Caption = "ADFS",
                        SignInAsAuthenticationType = signInAsType,
                        MetadataAddress = Settings.Federation.MetadataAddress,
                        Wtrealm = Settings.Federation.Identifier,
                        CallbackPath = new PathString(Settings.Federation.CallbackPath),
                        Wreply = "https://dev/identity/adfs",
                        SignOutWreply = "https://dev/identity/signoutcallback",
                        Notifications = new WsFederationAuthenticationNotifications
                        {
                            RedirectToIdentityProvider = notification =>
                            {
                                if (notification.ProtocolMessage.IsSignOutMessage)
                                {
                                    var signOutMessageId = notification.OwinContext.Environment.GetSignOutMessageId();
                                    if (signOutMessageId != null)
                                    {
                                        notification.OwinContext.Response.Cookies.Append("state", signOutMessageId);
                                    }
                                }

                                return Task.FromResult(0);
                            }
                        }
                    };
                    app.UseWsFederationAuthentication(adfs);
            }

**Comment by solidstore** [comment]
I added a call to ProcessFederatedSignoutAsync inside the wsignoutcleanup1.0 but noticed that the iframe it renders was never requested.


`<iframe src='https://dev/identity/connect/endsessioncallback?sid=973d30eb95deec9618c0ca1655ea40f9'></iframe>`

That turned out to be X-Frame-Options: SAMEORIGIN headers that had been added to the idserver responses. We that removed, the request is made.

The only issue now is that the wreply callback to /identity/signoutcallback is being hit, I am calling RenderLoggedOutViewAsync but the response ends with a 404.
